### PR TITLE
[improve][doc] improve Pulsar function document

### LINKED ISSUE #18593: [Doc] Pulsar function document need improve
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### What issue do you find in Pulsar docs?

Pulsar function document is hard to read for Pulsar beginners.

For example, many errors:

1. the wrong usage of `putstate`:

image

2. document for window functions is the same as normal functions.

And so on. 

### What is your suggestion?

Improve all the documents for Pulsar function to be easy-use.

### Any reference?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by labuladong**
Fixes #18593

### Modifications

Most of the changes are style improvements.

Some key changes are below:

1. Request users to set `functionsWorkerEnabled=true` before playing with Pulsar function.
2. For clarity, always use yaml file to create functions instead of CLI parameters, and remove the redundant `UserConfig` field in the example yaml file.
3. Rewrite the document for window function. The original document for window function is wrong.
4. Move `example-stateful-function-config.yaml` to `examples` folder when compile.

### Documentation

- [x] `doc` 
- [ ] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

**Comment by labuladong** [comment]
cc @tisonkun @Anonymitaet @momo-jun

**Comment by labuladong** [comment]
Sorry I found there still are some improvements, so mark this pr to draft for this moment.

File: site2/docs/functions-quickstart.md [hunk 1]
```diff
@@ -108,74 +108,114 @@ Before starting functions, you need to [start Pulsar](#start-standalone-pulsar).
    "test/test-namespace"
    ```
 
-3. In the same terminal window as step 1, create a function named `examples`.
+## Start functions
+
+:::note
+
+Before starting functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-tenantnamespace-for-test).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:115]:
```suggestion
Before starting functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-namespace-for-test).
```

File: site2/docs/functions-quickstart.md [hunk 2]
```diff
@@ -246,20 +286,16 @@ The following example provides instructions to validate counter functions.
 
 :::note
 
-Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar).
+Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-tenantnamespace-for-test).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:289]:
```suggestion
Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-namespace-for-test).
```

File: site2/docs/functions-quickstart.md [hunk 3]
```diff
@@ -268,34 +304,55 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
+   tenant: "test"
+   namespace: "test-namespace"
+   name: "word_count"
+   className: "org.apache.pulsar.functions.api.examples.WordCountFunction"
+   inputs: ["test_wordcount_src"] # this function will read messages from these topics
+   autoAck: true
+   parallelism: 1
+   ```
+   
+   You can see the source code of `WordCountFunction` [here](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/WordCountFunction.java). This function won't return any value but store the occurrence of words in function context. So we don't need to specify an output topic.
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:319]:
```suggestion
   You can see the [source code](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/WordCountFunction.java) of `WordCountFunction`. This function won't return any value but store the occurrence of words in function context. So you don't need to specify an output topic.
```

File: site2/docs/functions-quickstart.md [hunk 4]
```diff
@@ -268,34 +304,55 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
+   More information about the yaml config see [here](functions-cli.md#yaml-configurations-for-pulsar-functions).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:320]:
```suggestion
   For more information about the yaml config, see the [reference](functions-cli.md#yaml-configurations-for-pulsar-functions).
```

File: site2/docs/functions-quickstart.md [hunk 5]
```diff
@@ -414,110 +455,130 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
 ## Start window functions
 
 Window functions are a special form of Pulsar Functions. For more information, see [concepts](functions-concepts.md#window-function).
 
+The following example provides instructions to start a window function to calculate the sum in a window.
+
 :::note
 
-Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar).
+Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar)  and [create test namespace](#create-a-tenantnamespace-for-test).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:474]:
```suggestion
Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar)  and [create test namespace](#create-a-namespace-for-test).
```

File: site2/docs/functions-quickstart.md [hunk 6]
```diff
@@ -414,110 +455,130 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
+   
+   # every 5 messages, calculate sum of the latest 10 messages
+   windowConfig:
+     windowLengthCount: 10
+     slidingIntervalCount: 5
    ```
 
-   **Output**
+   You can see the source code of `AddWindowFunction` [here](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/AddWindowFunction.java).
+   More information about the yaml config see [here](functions-cli.md#yaml-configurations-for-pulsar-functions).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:511]:
```suggestion
   For more information about the yaml config, see the [reference](functions-cli.md#yaml-configurations-for-pulsar-functions).
```

File: site2/docs/functions-quickstart.md [hunk 7]
```diff
@@ -414,110 +455,130 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
+
 :::note
 
-Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar).
+Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar)  and [create test namespace](#create-a-tenantnamespace-for-test).
 
 :::
 
-1. Create a tenant and a namespace.
+1. create a function using `example-window-function-config.yaml`.
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:478]:
```suggestion
1. Create a function using `example-window-function-config.yaml`.
```

File: site2/docs/functions-quickstart.md [hunk 8]
```diff
@@ -108,74 +108,114 @@ Before starting functions, you need to [start Pulsar](#start-standalone-pulsar).
...
-3. In the same terminal window as step 1, create a function named `examples`.
+## Start functions
+
+:::note
+
+Before starting functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-tenantnamespace-for-test).
+
+:::
+
+1. create a function named `examples`.
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:119]:
```suggestion
1. Create a function named `examples`.
```

File: site2/docs/functions-quickstart.md [hunk 9]
```diff
@@ -108,74 +108,114 @@ Before starting functions, you need to [start Pulsar](#start-standalone-pulsar).
...
+   name: "example" # function name
+   className: "org.apache.pulsar.functions.api.examples.ExclamationFunction"
+   inputs: ["test_src"] # this function will read messages from these topics
+   output: "test_result" # the return value of this function will be sent to this topic  
+   autoAck: true # function will acknowledge input messages if set true 
+   parallelism: 1
+   ```
+
+   You can see the source code of `ExclamationFunction` [here](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/ExclamationFunction.java).
+   More information about the yaml config see [here](functions-cli.md#yaml-configurations-for-pulsar-functions).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:155]:
```suggestion
   For more information about the yaml config, see the [reference](functions-cli.md#yaml-configurations-for-pulsar-functions). 
```

File: site2/docs/functions-quickstart.md [hunk 10]
```diff
@@ -108,74 +108,114 @@ Before starting functions, you need to [start Pulsar](#start-standalone-pulsar).
...
+   namespace: "test-namespace"
+   name: "example" # function name
+   className: "org.apache.pulsar.functions.api.examples.ExclamationFunction"
+   inputs: ["test_src"] # this function will read messages from these topics
+   output: "test_result" # the return value of this function will be sent to this topic  
+   autoAck: true # function will acknowledge input messages if set true 
+   parallelism: 1
+   ```
+
+   You can see the source code of `ExclamationFunction` [here](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/ExclamationFunction.java).
```
Comment (momo-jun) [inline site2/docs/functions-quickstart.md:154]:
```suggestion
   You can see the [source code](https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/ExclamationFunction.java) of `ExclamationFunction`.
```

**Comment by momo-jun** [comment]
@labuladong thanks for improving the docs! Feel free to ping me for another look when it's ready.

**Comment by labuladong** [comment]
Change addressed, ready for review~

**Review by freeznet — APPROVED** [review]

File: site2/docs/functions-package-python.md [hunk 11]
```diff
@@ -16,16 +16,16 @@ To package a Python function into **one Python file**, complete the following st
 1. Write a Python function.
 
    ```python
-    from pulsar import Function #  import the Function module from Pulsar
-
-    # The classic ExclamationFunction that appends an exclamation at the end
-    # of the input
-    class ExclamationFunction(Function):
+   from pulsar import Function #  import the Function module from Pulsar
+
```
Comment (cbornet) [inline site2/docs/functions-package-python.md:20]: extra spaces ?

File: site2/docs/functions-package-python.md [hunk 12]
```diff
@@ -16,16 +16,16 @@ To package a Python function into **one Python file**, complete the following st
...
+   from pulsar import Function #  import the Function module from Pulsar
+   
+   # The classic ExclamationFunction that appends an exclamation at the end
+   # of the input
+   class ExclamationFunction(Function):
       def __init__(self):
-        pass
-
+         pass
+
```
Comment (cbornet) [inline site2/docs/functions-package-python.md:26]: extra spaces ?

File: site2/docs/functions-package-python.md [hunk 13]
```diff
@@ -39,27 +39,33 @@ To package a Python function into **one Python file**, complete the following st
 2. Install a Python client. The implementation of a Python function depends on the Python client. 
 
    ```bash
-    pip install pulsar-client==2.10.0
+   pip install pulsar-client==2.10.0
+   ```
+   
+   And install protobuf tools to generate proto file:
```
Comment (cbornet) [inline site2/docs/functions-package-python.md:45]:
```suggestion
   And install protobuf tools to generate the proto files:
```

File: site2/docs/functions-quickstart.md [hunk 14]
```diff
@@ -108,74 +108,114 @@ Before starting functions, you need to [start Pulsar](#start-standalone-pulsar).
    "test/test-namespace"
    ```
 
-3. In the same terminal window as step 1, create a function named `examples`.
+## Start functions
+
+:::note
+
+Before starting functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-namespace-for-test).
```
Comment (cbornet) [inline site2/docs/functions-quickstart.md:115]:
```suggestion
Before starting functions, you need to [start Pulsar](#start-standalone-pulsar) and [create a test namespace](#create-a-namespace-for-test).
```

File: site2/docs/functions-quickstart.md [hunk 15]
```diff
@@ -246,20 +286,16 @@ The following example provides instructions to validate counter functions.
 
 :::note
 
-Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar).
+Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar) and [create test namespace](#create-a-namespace-for-test).
```
Comment (cbornet) [inline site2/docs/functions-quickstart.md:289]:
```suggestion
Before starting stateful functions, you need to [start Pulsar](#start-standalone-pulsar) and [create a test namespace](#create-a-namespace-for-test).
```

File: site2/docs/functions-quickstart.md [hunk 16]
```diff
@@ -414,110 +455,130 @@ Before starting stateful functions, you need to [start Pulsar](#start-standalone
...
 ## Start window functions
 
 Window functions are a special form of Pulsar Functions. For more information, see [concepts](functions-concepts.md#window-function).
 
+The following example provides instructions to start a window function to calculate the sum in a window.
+
 :::note
 
-Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar).
+Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar)  and [create test namespace](#create-a-namespace-for-test).
```
Comment (cbornet) [inline site2/docs/functions-quickstart.md:474]:
```suggestion
Before starting window functions, you need to [start Pulsar](#start-standalone-pulsar)  and [create a test namespace](#create-a-namespace-for-test).
```

**Review by Anonymitaet — APPROVED** [review]

File: site2/docs/functions-package-python.md (hunk 12 above)
Comment (labuladong) [reply site2/docs/functions-package-python.md:26]: Yes, python default indent is 4 spaces, so I add extra spaces. I've tested it, users can copy & paste this code snippet to a file directly.

File: site2/docs/functions-package-python.md (hunk 11 above)
Comment (labuladong) [reply site2/docs/functions-package-python.md:20]: Same reason as below.

**Review by momo-jun — APPROVED** [review]

File: site2/docs/functions-package-python.md (hunk 12 above)
Comment (cbornet) [reply site2/docs/functions-package-python.md:26]: No, I mean there are spaces in L26 that I believe should not be there since the line is empty

Comment (cbornet) [inline site2/docs/functions-package-python.md:26]:
```suggestion

```

File: site2/docs/functions-package-python.md (hunk 11 above)
Comment (cbornet) [inline site2/docs/functions-package-python.md:20]:
```suggestion

```

Comment (cbornet) [reply site2/docs/functions-package-python.md:20]: See my comment below

File: site2/docs/functions-package-python.md (hunk 12 above)
Comment (labuladong) [reply site2/docs/functions-package-python.md]: Understand. I have removed them.

**Comment by labuladong** [comment]
Thanks for your suggestion~ PTAL @cbornet

**Review by cbornet — APPROVED** [review]

**Comment by cbornet** [comment]
LGTM 👍

**Comment by labuladong** [comment]
Can this PR be merged?

**Comment by tisonkun** [comment]
CI failed. Retrigger.

**Review by tisonkun — APPROVED** [review]

**Comment by Technoboy-** [comment]
Cherry picked some parts to branch-2.11.   https://github.com/apache/pulsar/pull/20073

**Pull request merged**